packages/notifier/tools/subscribe.js

```diff
@@ -0,0 +1,148 @@
+import { E, Far } from '@endo/far';
+
+import '../src/types-ambient.js';
+
+const sink = () => {};
+
+/**
+ * Create a near iterable that corresponds to a potentially far one.
+ *
+ * @template T
+ * @param {ERef<AsyncIterable<T>>} itP
+ * @returns {AsyncIterable<T>}
+ */
+export const subscribe = itP =>
+  Far('AsyncIterable', {
+    [Symbol.asyncIterator]: () => {
+      const it = E(itP)[Symbol.asyncIterator]();
+      return Far('AsyncIterator', {
+        next: async () => E(it).next(),
+      });
+    },
+  });
+
+/**
+ * Asyncronously iterates over the contents of a PublicationList as they appear.
+ * This iteration must drop parts of publication records that are no longer
+ * needed so they can be garbage collected.
+ *
+ * @template T
+ * @param {PublicationList<T>} pubList
+ * @returns {AsyncIterableIterator<T, T>}
+ */
+const makeEachIterator = pubList => {
+  // To understand the implementation, start with
+  // https://web.archive.org/web/20160404122250/http://wiki.ecmascript.org/doku.php?id=strawman:concurrency#infinite_queue
+  return Far('EachIterator', {
+    next: () => {
+      const resultP = E.get(pubList).head;
+      pubList = E.get(pubList).tail;
+      // We expect the tail to be the "cannot read past end" error at the end
+      // of the happy path.
+      // Since we are wrapping that error with eventual send, we sink the
+      // rejection here too so it doesn't become an invalid unhandled rejection
+      // later.
+      void E.when(pubList, sink, sink);
+      return resultP;
+    },
+    [Symbol.asyncIterator]: () => makeEachIterator(pubList),
+  });
+};
+
+/**
+ * Given a local or remote subscriber, returns a local AsyncIterable which
+ * provides "prefix lossy" iterations of the underlying PublicationList.
+ * By "prefix lossy", we mean that you may miss everything published before
+ * you ask the returned iterable for an iterator. But the returned iterator
+ * will enumerate each thing published from that iterator's starting point.
+ *
+ * If the underlying PublicationList is terminated, that terminal value will be
+ * reported losslessly.
+ *
+ * @template T
+ * @param {ERef<EachTopic<T>>} topic
+ */
+export const subscribeEach = topic => {
+  const iterable = Far('EachIterable', {
+    [Symbol.asyncIterator]: () => {
+      const pubList = E(topic).subscribeAfter();
+      return makeEachIterator(pubList);
+    },
+  });
+  return iterable;
+};
+harden(subscribeEach);
+
+/**
+ * @template T
+ * @param {ERef<LatestTopic<T>>} topic
+ * @param {UpdateCount} [localUpdateCount]
+ * @returns {AsyncIterableIterator<T>}
+ */
+const makeLatestIterator = (topic, localUpdateCount) => {
+  let myIterationResultP;
+  return Far('LatestIterator', {
+    [Symbol.asyncIterator]: () => makeLatestIterator(topic, localUpdateCount),
```

Comment:
Same comment as for `makeEachIterator` regarding this surprising behavior.

Reply:
Noted.